### DISCUSSION TITLE
automatic title generation for code-blocks

### DISCUSSION THREAD

**Discussion by meeusr** [mkdocs/mkdocs#3574] [Q&A]
Hi

we are using the material mkdocs to generate documentation of our project.
Next to that we have also other scripts that use the md files as input.

Below is an example of how the start of a code-block looks like in our documentation:
{ #link data-sw-release="2412" data-application="application_name" }
These "data" items are used by our private script alone for the moment but I would like to use these also to generate an appropriate header/title for the code-block. For example it could indicate on which application this codeblock applies.

Would that be possible in some way?

Thanks
Ronny

**Comment by pawamoy** [comment]
Does this line appear *within* the code block, or *above* the code block?

**Reply by meeusr** [reply]
On  the same line as the 3 backticks to start the codeblock, like below:
```
   ``` { #link data-sw-release="2412" data-application="application_name" }
```

**Reply by pawamoy** [reply]
Then I guess you could create a [custom superfence](https://facelessuser.github.io/pymdown-extensions/extensions/superfences/) to automatically define a title based on these attributes.

**Reply by meeusr** [reply]
When reading the documentation, it looks like indeed that is a valid option.
Thanks for this suggestion.

But to be honest: I have no clue how to add this to my project. The documentation is not clear to me :-(.
I thought of using the https://facelessuser.github.io/pymdown-extensions/extensions/superfences/#validators option to generate the title based on the passed attributes but I have no clue of how to specify this in the yaml file nor where to put the validator code ...

**Reply by oprypin** [reply]
@meeusr If you don't have confidence in implementing this stuff, you shouldn't ask your question in a shape that can only solicit general vague pointers.

What's needed is a set of example inputs, a set of precisely specified expected outputs, and ideally what's the reasoning why the outputs are like that based on the inputs.

**Reply by meeusr** [reply]
I'm very sorry to ask stupid questions, but before the answer of @pawamoy it was not clear to me at all of it was even possible ...

Basically my input is like specified in my first reply:

```
   ``` { #link data-sw-release="2412" data-application="application_name" }
```

this should be translated to:
```
    ``` { #link title="application_name release 2412" data-sw-release="2412" data-application="application_name" }
```

before generating the documentation.
I can understand that this can be done using the custom superfence extension but the only question is how this can be plugged in in the mkdocs document.

**Reply by pawamoy** [reply]
The only stupid questions are the ones that are not asked :slightly_smiling_face: (quote from... someone).

It *does* help though to provide relevant information, as stated by @oprypin, as it helps us giving better answers.

The way to configure your custom fence, is the same way you configure Mermaid diagrams for example. PyMDown's docs doesn't provide MkDocs configuration examples, because it is independent of MkDocs, but think it would help users if it did. It's likely used through MkDocs the most.

```yaml
markdown_extensions:
  - pymdownx.superfences:
      custom_fences:
        - name: your_fence_name
          class: your_class_name_or_empty
          format: !!python/name:path.to.your.formatter
```

More examples in one of my own projects: https://github.com/pawamoy/markdown-exec.
